### PR TITLE
[ONNX] Add support for tensordot operator.

### PR DESCRIPTION
**Pull request by fatcat-z**
Add the support for tensordot operator to exporter. It leverages einsum() operator to achieve the goal.
Also add 2 tests for this operator.

### CODE REVIEW

File: torch/onnx/symbolic_opset12.py [hunk 1]
```diff
@@ -209,3 +209,48 @@ def unfold(g, input, dimension, size, step):
...
         return _unimplemented("Unfold", "input size not accessible")
+
+@parse_args('v', 'v', 'is', 'is', 'v')
+def tensordot(g, input_a, input_b, dims_a, dims_b, out=None):
+    if out is not None:
+        _unimplemented("Tensordot", "Out parameter is not supported for tensordot.")
+
+    from torch.onnx.symbolic_opset9 import view, permute
+    for i in range(len(dims_a)):
+        if sym_help._get_tensor_dim_size(input_a, dims_a[i]) != sym_help._get_tensor_dim_size(input_b, dims_b[i]):
```
Comment (jiafatom) [inline torch/onnx/symbolic_opset12.py:220]: If both returns `None`, shall we also raise exception?

File: torch/onnx/symbolic_opset12.py [hunk 2]
```diff
@@ -209,3 +209,48 @@ def unfold(g, input, dimension, size, step):
...
+            raise RuntimeError("Dimensions are not matched for operator tensordot: a - {}, b - {}.".format(dims_a, dims_b))
+
+    dim_count_a = input_a.type().dim()
+    dim_count_b = input_b.type().dim()
+    for i in range(len(dims_a)):
```
Comment (jiafatom) [inline torch/onnx/symbolic_opset12.py:225]: Use list comprehension to write this?

File: torch/onnx/symbolic_opset12.py [hunk 3]
```diff
@@ -209,3 +209,48 @@ def unfold(g, input, dimension, size, step):
...
+        if (dims_a[i] < 0):
+            dims_a[i] += dim_count_a
+    for i in range(len(dims_b)):
```
Comment (jiafatom) [inline torch/onnx/symbolic_opset12.py:228]: list comprehension?

File: torch/onnx/symbolic_opset12.py [hunk 4]
```diff
@@ -209,3 +209,48 @@ def unfold(g, input, dimension, size, step):
...
+    left_dims_sizes_a = [sym_help._get_tensor_dim_size(input_a, left_dims_a[i])
+                         for i in range(len(left_dims_a))]
+    left_dims_sizes_b = [sym_help._get_tensor_dim_size(input_b, left_dims_b[i])
+                         for i in range(len(left_dims_b))]
+
+    new_input_a = permute(g, input_a, left_dims_a + dims_a)
+    new_input_b = permute(g, input_b, dims_b + left_dims_b)
+
+    output_a = view(g, new_input_a, left_dims_sizes_a + [-1])
+    output_a = view(g, output_a, [-1] + [sym_help._get_tensor_dim_size(output_a, len(left_dims_sizes_a))])
```
Comment (jiafatom) [inline torch/onnx/symbolic_opset12.py:248]: Can we directly use `view` once? Feel like the first `view` is only used to calculate the dim size.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 5]
```diff
@@ -7831,6 +7831,32 @@ def forward(self, x, window_length: int):
...
+    def test_tensordot_dim_count(self):
+        class M(torch.nn.Module):
+            def forward(self, x, y):
+                output = torch.tensordot(x, y, 2)
+                return output
+
+        x = torch.randint(6, (7, 5, 3, 4))
+        y = torch.randint(6, (3, 4, 8, 9))
+
+        self.run_test(M(), (x, y))
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:7845]: you can extend this test with dynamic input size (test with input of different sizes after export), by using argument `dynamic_axes` and `test_with_inputs`

File: torch/onnx/symbolic_opset12.py [hunk 6]
```diff
@@ -209,3 +209,48 @@ def unfold(g, input, dimension, size, step):
...
+def tensordot(g, input_a, input_b, dims_a, dims_b, out=None):
+    if out is not None:
+        _unimplemented("Tensordot", "Out parameter is not supported for tensordot.")
+
+    from torch.onnx.symbolic_opset9 import view, permute
+    for i in range(len(dims_a)):
+        if sym_help._get_tensor_dim_size(input_a, dims_a[i]) != sym_help._get_tensor_dim_size(input_b, dims_b[i]):
+            raise RuntimeError("Dimensions are not matched for operator tensordot: a - {}, b - {}.".format(dims_a, dims_b))
+
+    dim_count_a = input_a.type().dim()
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset12.py:223]: I think there is helper function for getting rank in `sym_help` as well.

Also for this implementation, it would seem the tensor rank is required to be known as static number. This comes from limitation of potential `transpose/permute` later, which requires static rank, as well as from the code below for constructing dimensions. In that case, it would be better to throw not supported error when rank is unknown. (However, if `transpose` is not used, then we can lift the requirement on rank, by doing the below construction on dynamic shape with nodes of `ONNX::Shape`, and `ONNX::Slice`)

File: torch/onnx/symbolic_opset12.py [hunk 7]
```diff
@@ -209,3 +209,48 @@ def unfold(g, input, dimension, size, step):
...
+            dims_b[i] += dim_count_b
+
+    left_dims_a = [i
+                   for i in range(dim_count_a)
+                   if (i not in dims_a)]
+    left_dims_b = [i
+                   for i in range(dim_count_b)
+                   if (i not in dims_b)]
+
+    left_dims_sizes_a = [sym_help._get_tensor_dim_size(input_a, left_dims_a[i])
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset12.py:239]: similar to David's comment, need to do None checks for returned values of `_get_tensor_dim_size`

File: torch/onnx/symbolic_opset12.py (hunk 1 above)
Comment (fatcat-z) [reply torch/onnx/symbolic_opset12.py:220]: Good catch.

File: torch/onnx/symbolic_opset12.py (hunk 6 above)
Comment (fatcat-z) [reply torch/onnx/symbolic_opset12.py]: Update code to infer the size.

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 5 above)
Comment (fatcat-z) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:7963]: Add one more test for this.

File: torch/onnx/symbolic_opset12.py (hunk 4 above)
Comment (fatcat-z) [reply torch/onnx/symbolic_opset12.py]: In my latest commit, view() has been removed. Please help to take another review. Thanks

File: torch/onnx/symbolic_opset12.py (hunk 1 above)
Comment (fatcat-z) [reply torch/onnx/symbolic_opset12.py]: Has removed the check for the size here. Because sym_help._get_tensor_dim_size() will probably return a None in a reasonable scenario, like in my newly added test which uses dynamic_axes.

In that case, the size of those dimensions could not be got by calling sym_help._get_tensor_dim_size() method, but with my latest changes, we don't need to get the size values so None will has no impact to the function at all.

File: torch/csrc/jit/passes/erase_number_types.cpp [hunk 8]
```diff
@@ -6,13 +6,19 @@
...
+}
+
 void EraseNumberTypesOnBlock(Block* block) {
   for (auto it = block->nodes().begin(), end = block->nodes().end(); it != end;
        ++it) {
     for (auto inp : it->inputs()) {
-      if (inp->type()->isSubtypeOf(NumberType::get())) {
-        inp->setType(TensorType::get());
-      }
+      SetNumTypeToTensorType(inp);
```
Comment (jiafatom) [inline torch/csrc/jit/passes/erase_number_types.cpp:21]: We originally only deal with NumberType not BoolType, now it considers BoolType as well. Is this on purpose?

File: torch/onnx/symbolic_opset9.py [hunk 9]
```diff
@@ -67,6 +67,8 @@ def _shape_as_tensor(g, input):
 
 
 def _reshape_from_tensor(g, input, shape):
+    if (type(shape) is list):
```
Comment (jiafatom) [inline torch/onnx/symbolic_opset9.py:70]: isinstance(shape, list)?
